Unify the dtype to VecMask<float, N> in ops.masked

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126662

**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/126449. For `ops.masked` in CPP backend, when input dtype is `bool`, we actually load it as `VecMask<float, N>`. So, we should unify the type of `other` and `mask` to the same as  `VecMask<float, N>` to invoke `blendv` method.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_ops_masked_with_bool_input
clear && PYTORCH_ALL_SAMPLES=1 python -u -m pytest -s -v test/inductor/test_torchinductor_opinfo.py -k test_comprehensive__chunk_cat_cpu_bool
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
Unify the dtype to VecMask<float, N> in ops.masked

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Unify the dtype to VecMask<float, N> in ops.masked"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Unify the dtype to VecMask<float, N> in ops.masked"


**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/126449. For `ops.masked` in CPP backend, when input dtype is `bool`, we actually load it as `VecMask<float, N>`. So, we should unify the type of `other` and `mask` to the same as  `VecMask<float, N>` to invoke `blendv` method.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_ops_masked_with_bool_input
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

isuruf reviewed (approved):
Looks good. Thanks

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1419,6 +1419,12 @@ def masked(mask, body, other):
             with V.kernel.swap_buffers(code), code.indent():
+                if dtype == torch.bool and result.is_vec:
+                    # Unify blendv to use the VecMask<float, N>, since loading bool as VecMask<float> by default
+                    dtype = torch.float
+                    num_vectors = V.kernel._get_num_vectors(dtype)
+                    other_code_vec = f"at::vec::VecMask<{DTYPE_TO_CPP[dtype]}, {num_vectors}>::from({other_code})"
```
simplify this with `_get_mask_type`?

leslie-fang-intel replied:
Thanks. Simplified with `_get_mask_type`.

peterbell10 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1419,6 +1419,12 @@ def masked(mask, body, other):
             code = BracesBuffer()
             code.writeline("[&]")
             with V.kernel.swap_buffers(code), code.indent():
```
NIT: It looks like this is re-implementing `ops.where`. Could this whole block just be `self.overrides.where(new_mask, body_code_vec, other_code_vec)` or similar?

jgong5 replied:
That's good suggestion. I will let @leslie-fang-intel to decide. :)

leslie-fang-intel replied:
Thanks for the suggestions. Create the corresponding `CSEVariable` to reuse `V.kernel.overrides.where`. Let's see what's CI think about.

leslie-fang-intel committed:
Update on "Unify the dtype to VecMask<float, N> in ops.masked"


**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/126449. For `ops.masked` in CPP backend, when input dtype is `bool`, we actually load it as `VecMask<float, N>`. So, we should unify the type of `other` and `mask` to the same as  `VecMask<float, N>` to invoke `blendv` method.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_ops_masked_with_bool_input
clear && PYTORCH_ALL_SAMPLES=1 python -u -m pytest -s -v test/inductor/test_torchinductor_opinfo.py -k test_comprehensive__chunk_cat_cpu_bool
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Unify the dtype to VecMask<float, N> in ops.masked"


**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/126449. For `ops.masked` in CPP backend, when input dtype is `bool`, we actually load it as `VecMask<float, N>`. So, we should unify the type of `other` and `mask` to the same as  `VecMask<float, N>` to invoke `blendv` method.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_ops_masked_with_bool_input
clear && PYTORCH_ALL_SAMPLES=1 python -u -m pytest -s -v test/inductor/test_torchinductor_opinfo.py -k test_comprehensive__chunk_cat_cpu_bool
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

isuruf commented:
Can you add `_chunk_cat` and `constant_pad_nd` to the list at https://github.com/pytorch/pytorch/blob/51c07f9f69aedf884fc697f3ef8545cb0303e2a9/test/inductor/test_torchinductor_opinfo.py#L410 ?

leslie-fang-intel commented:
> Can you add `_chunk_cat` and `constant_pad_nd` to the list at
> 
> https://github.com/pytorch/pytorch/blob/51c07f9f69aedf884fc697f3ef8545cb0303e2a9/test/inductor/test_torchinductor_opinfo.py#L410
> 
> ?

Sure, added.

leslie-fang-intel committed:
Update on "Unify the dtype to VecMask<float, N> in ops.masked"


**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/126449. For `ops.masked` in CPP backend, when input dtype is `bool`, we actually load it as `VecMask<float, N>`. So, we should unify the type of `other` and `mask` to the same as  `VecMask<float, N>` to invoke `blendv` method.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_ops_masked_with_bool_input
clear && PYTORCH_ALL_SAMPLES=1 python -u -m pytest -s -v test/inductor/test_torchinductor_opinfo.py -k test_comprehensive__chunk_cat_cpu_bool
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Unify the dtype to VecMask<float, N> in ops.masked"


**Summary**
Fix issue: https://github.com/pytorch/pytorch/issues/126449. For `ops.masked` in CPP backend, when input dtype is `bool`, we actually load it as `VecMask<float, N>`. So, we should unify the type of `other` and `mask` to the same as  `VecMask<float, N>` to invoke `blendv` method.

**Test Plan**
```
clear && python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_ops_masked_with_bool_input
clear && PYTORCH_ALL_SAMPLES=1 python -u -m pytest -s -v test/inductor/test_torchinductor_opinfo.py -k test_comprehensive__chunk_cat_cpu_bool
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -r

leslie-fang-intel committed:
Update

[ghstack-poisoned]

isuruf commented:
@pytorchbot merge
